android/gltfio-android/src/main/cpp/AssetLoader.cpp

```diff
@@ -23,18 +23,151 @@
 
 #include <gltfio/AssetLoader.h>
 #include <gltfio/MaterialProvider.h>
+#include <utils/debug.h>
 
 #include "common/NioUtils.h"
 
+#include "MaterialKey.h"
+
 using namespace filament;
 using namespace gltfio;
 using namespace utils;
 
+class JavaMaterialProvider : public MaterialProvider {
+    JNIEnv* const mEnv;
+    const jobject mJavaProvider;
+    mutable Material** mPreviousMaterials = nullptr;
+
+    jclass mMaterialKeyClass;
+
+    jmethodID mMaterialKeyConstructor;
+    jmethodID mCreateMaterialInstance;
+    jmethodID mGetMaterials;
+    jmethodID mNeedsDummyData;
+    jmethodID mDestroyMaterials;
+    jmethodID mMaterialInstanceGetNativeObject;
+    jmethodID mMaterialGetNativeObject;
+
+public:
+    JavaMaterialProvider(JNIEnv* env, jobject provider) : mEnv(env),
+            mJavaProvider(env->NewGlobalRef(provider)) {
+        mMaterialKeyClass = env->FindClass(JAVA_MATERIAL_KEY);
+        mMaterialKeyClass = (jclass) env->NewGlobalRef(mMaterialKeyClass);
+        assert_invariant(mMaterialKeyClass);
+
+        mMaterialKeyConstructor = env->GetMethodID(mMaterialKeyClass, "<init>", "()V");
+        assert_invariant(mMaterialKeyConstructor);
+
+        jclass materialInstanceClass = env->FindClass("com/google/android/filament/MaterialInstance");
```

Comment:
Just as a heads up, like with https://github.com/google/filament/commit/0a6a74c78ce18d3ad03a22c9aee408ce67f4f5ea, we'll have to cast the string literals to `char*`if this ever ends up in Google3.

Reply:
oh god. good catch.

Reply:
Luckily I think I'm in the clear, these env methods seem to take `const char*` rather than `char*`.

Reply:
I think the latest JDK correctly has them take a `const char*`, but Google3's version (for whatever reason) takes a `char*`.

Reply:
Gah, okay, thanks for the heads up.